prioQ fixes for mon_mitm, little improvements, logging

crhbetz opened the pull request:
fixes https://github.com/Map-A-Droid/MAD/issues/579 by not dropping the prio q on recalculation for mon_mitm, but also reducing recalculation to one per 59 minutes,
fixes https://github.com/Map-A-Droid/MAD/issues/596 by ditching the complicated database query and just fetching all spawns for the next hour instead.

Both of these changes are feasible, because once all spawn / despawn times are known, the prioQ for mon_mitm actually just becomes a line sorting the spawn points by their spawn times. No actual prioQ flexibility is needed in this mode. It should lead to more robust and humanly understandable scanning behaviour. Also, I added a little code to check each prioQ event against the `remove_from_queue_backlog` threshold, because otherwise this would only happen during prioQ recalculation. It makes more sense this way imo, but it is also needed with hourly recalculations only, or the prioQ might fall very far behind. If an event is over the threshold, a loglevel warning message will be logged.

I limited changes of behaviour to mon_mitm mode for now because it might have unintended side effects on other modes.

I also added a couple of log lines and a little detail to existing lines to make it more understandable what is happening with prioQ.

sn0opy commented:
Could you give us a short explanation what to look for when testing this?

crhbetz commented:
It's a bit tricky, because right now there's no comparable logging at the points where I made changes.

There should be an increase in scanning performance especially because of the fix of #596 , but it would still be hard to read from statistics because when the prioQ would get few events, the workers would resume walking the normal route and still scan stuff. It should provide better / more stable time left on scanned mons especially leading up to the half hour mark, but that's again difficult to track.

The best you can probably do is compare the prioQ on the map (or the printed one from loglevel debug) with the spawn times of the points it wants to check, to see if the Q remains consistent - especially to confirm everything going right across the half hour mark and change of hours.

One might set a really aggressive `remove_from_queue_backlog` for a bit to see if that new part correctly skips overdue events.

ChaosEmpire commented:
Remove any events in priority queue that have been due for scanning before NOW - given time in seconds (Default: 0)

I understand that default 0 means , disable removing of entries in backlog
but i get lots of remove backlog entries

assuming that 300 means, remove anything older than 300 seconds from backlog, so a 30min spawn can max get find as 25min its possible that 0 means reove directly all backlog, but then we need to set high values to workaround removal of backlog entries, so it tries to do all work

btw the part min_time_left_seconds could be used in similiar way for the backlog
if we dont want to report pokemon with less than x seconds..we also dont plan a route for them and can remove them from backlog, even if backlog removal is disabled

from my understanding backlog removal makes only sense in a range of 1sec after event to min_time_left_seconds value

maybe this handling can be checked too

ChaosEmpire commented:
During a round based mon mitm, a 2nd device has the chance to encounter iv missing pokemon
how about using "idle" times to handle not encountered pokemon in a 2nd hop
something like mon mitm device pool does interact like an iv mitm scanner device and cleans the iv mitm queue that way
if you have enough devices, you will have sometimes a few seconds without a jump, in this moment a device could encounter pokeon nr3/4 of a cluster, this will increase the amount of iv scanned pokemon, instead of let the devices have "idle" moments, this could make a only iv mitm scanner obsolete and device usage more efficient, as sometimes iv mitm devices stand around for minutes
also it would increase the max amount of pokemon on the map, if all device could be used as a combined mon mitm/iv mitm device

crhbetz commented:
Thanks for your input @ChaosEmpire - you brought up some valid points.

1. regarding `min_time_left_seconds` - as I just confirmed in the code, this will only be used in `iv_mitm` mode at all. Additionally, the prioQ build process in `mon_mitm` only fetches future spawn events from the database, so the value would be irrelevant anyway. (after 1 hour when the next calculation happens, older events are not automatically discarded - the newly built Q will be appended and the older events would be handled by `remove_from_queue_backlog` if necessary).

2. regarding remove_from_queue_backlog: `assuming that 300 means, remove anything older than 300 seconds from backlog, so a 30min spawn can max get find as 25min its possible that 0 means reove directly all backlog, but then we need to set high values to workaround removal of backlog entries, so it tries to do all work`
You're correct that with my changes, the default value of 0 makes no sense, because it will indeed remove any event that's older than "now". However, you probably don't want to disable removing old events completely, because in situations like partial outages of your devices you could end up with a Q that's behind hours and hours and in essence jumping around aimlessly.
I think the proper way to handle it would be a sane default value like 900 seconds (15 minutes). I'd have to check if this would negatively / counter-intuitively impact other scan modes, though. Also, the description of `remove_from_queue_backlog` could be improved.
I will check this and submit a change regarding this.

3. mon_mitm + iv_mitm "collaboration" / "merge"
I understand your proposal and where you identify the possible advantage, however this is beyond both the scope of this PR and my current grasp of the code base. I think this could only be achieved in a new scan mode / a merge of mon_mitm and iv_mitm.
Further, if your mon_mitm workers already have downtimes, I think the more appropriate solution would be to limit event clustering to a maximum of 2 events (something I've already thought about making configurable too), so that when running smooth, there should never be more than 2 encounters available on any given event, thus "guaranteeing" a 100% encounter rate.
I will think about and look into the latter option.

crhbetz commented:
@ChaosEmpire 🏓 

- Added `max_clustering` option to `mon_mitm` settings. Most probable use case would be to set it to `2`, so that a maximum of two mon events get clustered and thus in theory - provided your devices can keep up at all times! - all mons should get IV checked (catering to the two mons per location / GMO limitation of PogoDroid). Default is no limit / the value of `max_coords_within_radius`.
Search for the line containing `clustered new events` in logs to see the result of this. On my areas, I get an increase of 30-35% in clustered events from setting this to two.
- removed `min_time_left_seconds` from `mon_mitm` settings (not removed from DB yet)
- `remove_from_queue_backlog` now defaults to `900`, reasoning is detailed in my previous comment

ChaosEmpire commented:
I tested the PR based on max clustering unlimited, 2 and 3 for several areas i m running in mon mitm mode

one area is surley in a very good relation of device amount versus spawns
the others are mixed to very bad density of devices

result of testing, lots of more pokemon on the map compared to original prioq, 
times surely better compared to mode without prioq, but with cluster 2 its nearly the same of amount of pokemon and iv scanned, with max cluster 3 more pokemon, without big change of iv scanned amount, but still with much better times, than round based, so for the first time after month i would say, its worth to use prioq
the area with few devices and much more spawnpoints profits the most, amount of pokemon increased between 60 and 80 percent, , times got better, and iv checked pokemon increaed,compared to round based and also old prioq was less good than this

From my point of view, better logging, better scanning...
good pr, since last changes more customizable

Seandals commented:
I have also been testing this PR for the past several days and have noticed a significant improvement using prioq in a large area with a large number of spawn points. No issues. This is an awesome update!
